components/resource_control/src/service.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{sync::Arc, time::Duration};
+
+use futures::StreamExt;
+use kvproto::{pdpb::EventType, resource_manager::ResourceGroup};
+use pd_client::{PdClient, Result, RpcClient, RESOURCE_CONTROL_CONFIG_PATH};
+use tikv_util::{box_err, error};
+
+use crate::ResourceGroupManager;
+
+#[derive(Clone)]
+pub struct ResourceManagerService {
+    manager: Arc<ResourceGroupManager>,
+    pd_client: Arc<RpcClient>,
+    // record watch revision
+    revision: i64,
+}
+
+impl ResourceManagerService {
+    /// Constructs a new `Service` with `ResourceGroupManager` and a `RpcClient`
+    pub fn new(
+        manager: Arc<ResourceGroupManager>,
+        pd_client: Arc<RpcClient>,
+    ) -> ResourceManagerService {
+        ResourceManagerService {
+            pd_client,
+            manager,
+            revision: 0,
+        }
+    }
+}
+
+impl ResourceManagerService {
+    pub async fn watch_resource_groups(&mut self) {
+        // Firstly, load all resource groups as of now.
+        match self.list_resource_groups() {
+            Ok((groups, revision)) => {
+                self.revision = revision;
+                for group in groups {
+                    self.manager.add_resource_group(group);
```

Comment:
I think it's acceptable to allow a short period of time that the resource groups are not inited, in the duration, some requests are fallback to the default group, this may affect the request latency a little(priority queue fallback to FIFO), but this impact should be very small and the duration is not very long.

Reply:
I think this does not matter because resource manager's init is after the server start, there can be a short period of time that has not been initialized. At this time will fall back to the default group, which should be acceptable. cc @glorv